cpp/src/reductions/scan/ewm.cu

```diff
@@ -0,0 +1,374 @@
+    if (input.has_nulls()) {
+      return null_roll_up(input, stream);
+    } else {
+      return rmm::device_uvector<cudf::size_type>(input.size(), stream);
+    }
+  }();
+  // denominators are all 1 and do not need to be computed
+  // pairs are all (beta, 1-beta x_i) except for the first one
+
+  if (!input.has_nulls()) {
+    auto data = thrust::make_zip_iterator(
+      thrust::make_tuple(input.begin<T>(), thrust::make_counting_iterator<size_type>(0)));
+    thrust::transform_inclusive_scan(rmm::exec_policy(stream),
+                                     data,
+                                     data + input.size(),
+                                     pairs.begin(),
+                                     ewma_noadjust_functor<T, false>{beta},
+                                     recurrence_functor<T>{});
+
+  } else {
+    /*
+    In this case, a denominator actually has to be computed. The formula is
+    y_{i+1} = (1 - alpha)x_{i-1} + alpha x_i, but really there is a "denominator"
+    which is the sum of the weights: alpha + (1 - alpha) == 1. If a null is
+    encountered, that means that the "previous" value is downweighted by a
+    factor (for each missing value). For example this would y_2 be for one null:
+    data = {x_0, NULL, x_1},
+    y_2 = (1 - alpha)**2 x_0 + alpha * x_2 / (alpha + (1-alpha)**2)
+
+    As such, the pairs must be updated before summing like the adjusted case to
+    properly downweight the previous values. But now but we also need to compute
+    the normalization factors and divide the results into them at the end.
+    */
+    auto device_view = column_device_view::create(input);
+    auto valid_it    = detail::make_validity_iterator(*device_view);
+
+    auto data = thrust::make_zip_iterator(thrust::make_tuple(
+      input.begin<T>(), thrust::make_counting_iterator<size_type>(0), valid_it, nullcnt.begin()));
+
+    thrust::transform_inclusive_scan(rmm::exec_policy(stream),
+                                     data,
+                                     data + input.size(),
+                                     pairs.begin(),
+                                     ewma_noadjust_functor<T, true>{beta},
+                                     recurrence_functor<T>());
+  }
+
+  // copy the second elements to the output for now
+  thrust::transform(rmm::exec_policy(stream),
+                    pairs.begin(),
+                    pairs.end(),
+                    output.begin(),
+                    [] __device__(pair_type<T> pair) -> T { return pair.second; });
+  return output;
+}
+
+template <typename T>
+std::unique_ptr<column> ewma(scan_aggregation const& agg,
+                             column_view const& input,
+                             rmm::cuda_stream_view stream,
+                             rmm::device_async_resource_ref mr)
+{
+  auto ewma_agg = (dynamic_cast<ewma_aggregation const*>(&agg));
+  if (ewma_agg == NULL) { CUDF_FAIL("Expected an EWMA aggregation."); }
```

Comment:
Is this actually reachable? A `CUDF_FAIL` here seems out of place because you would only get to this point if other libcudf code (the dispatcher) incorrectly forwarded the wrong kind of aggregation, right? I think we can remove this line.